scripts/link-checker/check-links.js

```diff
@@ -0,0 +1,364 @@
+    const source = result.base.resolved;
+    const destination = result.url.resolved;
+
+    if (result.broken) {
+        const reason = result.brokenReason;
+
+        addLink(source, destination, reason, brokenLinks);
+
+        // Always log broken links to the console.
+        logLink(source, destination, reason);
+
+    } else if (process.env.DEBUG) {
+
+        // Log successes when DEBUG is truthy.
+        logLink(source, destination, result.http.response.statusCode);
+    }
+}
+
+// Handles BLC 'page' events.
+function onPage(error, pageURL, brokenLinks) {
+    if (error) {
+        addLink(pageURL, pageURL, error.message, brokenLinks);
+        logLink(pageURL, pageURL, error.message);
+    }
+    else if (process.env.DEBUG) {
+        logLink(pageURL, pageURL, "PAGE_OK");
+    }
+}
+
+// Handles the BLC 'complete' event, which is raised at the end of a run.
+async function onComplete(brokenLinks) {
+    const filtered = excludeAcceptable(brokenLinks);
+
+    if (filtered.length > 0) {
+
+        // If we failed and a retry count was provided, retry. Note that retry count !==
+        // run count, so a retry count of 1 means run once, then retry once, which means a
+        // total run count of two.
+        if (maxRetries > 0 && retryCount < maxRetries) {
+            retryCount += 1;
+            console.log(`Retrying (${retryCount} of ${maxRetries})...`);
+            checkLinks();
+            return;
+        }
+
+        const list = filtered
+            .map(link => `:link: <${link.source}|${new URL(link.source).pathname}> → ${link.destination} (${link.reason})`)
+            .join("\n");
+
+        // Post the results to Slack.
+        await postToSlack("docs-ops", list);
+    }
+}
+
+/**
+    We exclude some links:
+    - Our generated API docs have lots of broken links.
+    - Our available versions page includes links to private repos.
+    - GitHub Edit Links may be broken, because the page might not yet exist!
+    - Our LinkedIn page, for some reason, returns an HTTP error (despite being valid).
+    - Our Visual Studio Marketplace link for the Azure Pipelines task extension,
+      although valid and publicly available, is reported as a broken link.
+    - A number of synthetic illustrative links come from our examples/tutorials.
+    - GitLab 503s for requests for protected pages that don't contain certain cookies.
+*/
+function getDefaultExcludedKeywords() {
+    return [
+        "/registry/packages/*/api-docs",
+        "https://api.pulumi.com/",
+        "https://github.com/pulls?",
+        "https://github.com/pulumi/docs/edit/master",
+        "https://github.com/pulumi/docs/issues/new",
+        "https://github.com/pulumi/registry/edit/master",
+        "https://github.com/pulumi/registry/issues/new",
+        "https://github.com/pulumi/pulumi-hugo/edit/master",
+        "https://github.com/pulumi/pulumi-hugo/issues/new",
+        "https://www.linkedin.com/",
+        "https://linkedin.com/",
```

Comment:
we _could_ probably prune out this list if we really wanted, but not sure which ones specifically are linked from docs and which are linked from registry pages, so I just left most of them in.

Reply:
Yeah, we can do this later. Good callout though.